### PR TITLE
Generic AWS CloudWatch Event Parser

### PR DESCRIPTION
**Pull request by adrianosela**
## Background 

When an AWS CloudWatch "event rule" is triggered, the data that CloudWatch sends to the rule target (e.g. a Lambda function, or a Kinesis Stream) is wrapped around some generic fields common to all CloudWatch events:

```
{
    "version": "0",
    "id": "cd2d702e-ab31-411b-9344-793ce56b1bc7",
    "detail-type": "{SOURCE_SPECIFIC_DETAIL_TYPE}",
    "source": "aws.{SOURCE_SERVICE}",
    "account": "111122223333",
    "time": "1970-01-01T00:00:00Z",
    "region": "us-east-1",
    "resources": [],
    "detail": {SOURCE_SERVICE_SPECIFIC_DETAIL_JSON}
}
```

In the future... GuardDuty Findings (and perhaps other security/auditing events) will arrive at the pipeline through a CloudWatch event rule with a Kinesis stream as the target of the rule. Thus we know that we will need to parse out the CloudWatch fields "wrapper" ...which leads to this PR.

## In This PR:

* CloudWatch Event JSON model class
* CloudWatch Event parser
* A small CloudWatch pipeline which at the moment only prints out the region of the events received as a first-iteration/test + some utility functions/classes only used for this pipeline
* The necessary changes to Payload.java and Parser.java to allow for pipeline Events to take on the new payload format

## Testing:

```
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.mozilla.secops.parser.ParserTest
[INFO] Tests run: 50, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.764 s - in com.mozilla.secops.parser.ParserTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 50, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  16.000 s
[INFO] Finished at: 2019-06-12T23:34:08Z
[INFO] ------------------------------------------------------------------------
```

```
16:35 $ mvn compile exec:java -Dexec.mainClass=com.mozilla.secops.cloudwatch.CloudWatch -Dexec.args='--inputFile=./target/classes/guardduty-sample-findings.txt --monitoredResourceIndicator=aws'

us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
us-west-2
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.460 s
[INFO] Finished at: 2019-06-12T16:35:25-07:00
[INFO] ------------------------------------------------------------------------
```

### CODE REVIEW

File: src/main/java/com/mozilla/secops/cloudwatch/CloudWatch.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
+package com.mozilla.secops.cloudwatch;
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/cloudwatch/CloudWatch.java]: Let's pull all the pipeline related code out of this PR and focus just on the parser for now, this part will likely look a bit different once we get there.

File: src/main/java/com/mozilla/secops/parser/CloudWatch.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package com.mozilla.secops.parser;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.mozilla.secops.parser.models.aws.cloudwatch.CloudWatchEvent;
+import com.mozilla.secops.parser.models.aws.guardduty.GuardDutyFinding;
+import java.io.Serializable;
+
+/** Payload parser for AWS CloudWatch Event data */
+public class CloudWatch extends PayloadBase implements Serializable {
```
Comment (ameihm0912) [inline src/main/java/com/mozilla/secops/parser/CloudWatch.java]: If the CloudWatch side of things is always going to simply be an encapsulation layer, lets treat it as such in the parser and just push the part we are interested in (e.g., the detail) down to the payload parser.

As an example, see the following:

https://github.com/mozilla-services/foxsec-pipeline/blob/ce21dcff76d4bc5b0ed7b1ac0905de9a58db186e/src/main/java/com/mozilla/secops/parser/Parser.java#L296-L305

You could do something like this:

* Add a new function, something like `stripCloudwatchEncapsulation`

* In that function, if it is a CloudWatch event, grab the `source` field and set it in the parent event, e.g.:

```java
e.setCloudWatchSource(cw.getSource());
```

* What would be returned would simply be what is in `detail`, and the payload parser can reference the parent event source field to determine if it should parse it

* Then you would just have a `GuardDuty` event type

**Comment by ameihm0912** [comment]
https://github.com/aws/aws-lambda-java-libs/blob/master/aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/ScheduledEvent.java seems to be the closest thing I can find to an object representation of the encapsulation type; that may be usable as well instead of needing to redefine the CloudWatch encapsulation as a model here.

File: src/main/java/com/mozilla/secops/parser/models/aws/guardduty/GuardDutyFinding.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
+package com.mozilla.secops.parser.models.aws.guardduty;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.io.Serializable;
+
+/** Describes the format of an AWS GuardDuty Finding */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class GuardDutyFinding implements Serializable {
```
Comment (ajvb) [inline src/main/java/com/mozilla/secops/parser/models/aws/guardduty/GuardDutyFinding.java]: it looks like the aws java sdk provides a guard duty finding model - https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/guardduty/model/Finding.html

Comment (adrianosela) [reply src/main/java/com/mozilla/secops/parser/models/aws/guardduty/GuardDutyFinding.java]: oh wow thanks!

**Comment by adrianosela** [comment]
closed in favor of https://github.com/mozilla-services/foxsec-pipeline/pull/158

**Pull request closed without merging**